docs/config.md
# Compile-Time Configuration {#config}

This page lists the public compile-time configuration API

| Define | Default | Valid Range | Description |
|--------|---------|-------------|-------------|
| \ref UDS_SYS | auto-detected | \ref uds_sys_ | target system |
| UDS_CUSTOM_MILLIS | unset | set or unset | if unset (default), iso14229 provides UDSMillis() for the detected @ref UDS_SYS platform. If set, the library user must provide their own UDSMillis() implementation. |
| UDS_TP_ISOTP_SOCK | unset | set or unset | builds transport: Linux kernel ISO-TP socket. Suitable for linux newer than 5.10. See \ref UDSServerTpIsoTpSockInit and \ref UDSClientTpIsoTpSockInit. |
| UDS_TP_ISOTP_C_SOCKETCAN | unset | set or unset | builds transport: isotp-c over SocketCAN. Suitable for Linux newer than 2.6.25. See \ref UDSServerTpISOTpCSocketCANInit and \ref UDSClientTpISOTpCSocketCANInit. |
| UDS_TP_ISOTP_C | unset* | set or unset | builds transport: isotp-c. Suitable for *everything*, but you must bring your own CAN interface. See \ref UDSServerTpISOTpCInit, \ref UDSClientTpISOTpCInit and the \ref porting-guide. |
| \ref UDS_LOG_LEVEL | \ref UDS_LOG_NONE | \ref uds_log_level_ | iso14229 internal log level. Set UDS_LOG_LEVEL=UDS_LOG_DEBUG for a pleasant first-time bringup experience, then turn it off when you're finished. |
| \ref UDS_SERVER_DEFAULT_P2_MS | 50 | - | Default P2 timeout (ms) |
| \ref UDS_SERVER_DEFAULT_P2_STAR_MS | 5000 | - | Default P2* timeout (ms) |
| `UDS_SERVER_DEFAULT_S3_MS` | 5100 | - | Session timeout (ms) |
| `UDS_SERVER_DEFAULT_POWER_DOWN_TIME_MS` | 60 | - | Delay before ECU reset (ms) |
| `UDS_SERVER_0x27_BRUTE_FORCE_MITIGATION_BOOT_DELAY_MS` | 1000 | - | Boot delay for security access (ms) |
| `UDS_SERVER_0x27_BRUTE_FORCE_MITIGATION_AUTH_FAIL_DELAY_MS` | 1000 | - | Delay after auth failure (ms) |
| `UDS_SERVER_SEND_BUF_SIZE` | 4095 | - | Send buffer size |
| `UDS_SERVER_RECV_BUF_SIZE` | 4095 | - |  Receive buffer size |
| `UDS_CLIENT_DEFAULT_P2_MS` | 150 | - | Default P2 timeout (ms) |
| `UDS_CLIENT_DEFAULT_P2_STAR_MS` | 1500  | - | Default P2* timeout (ms) |
| `UDS_CLIENT_SEND_BUF_SIZE` | 4095 | - | Send buffer size |
| `UDS_CLIENT_RECV_BUF_SIZE` | 4095 | - | Receive buffer size |

- *except on `UDS_SYS=UDS_SYS_ARDUINO`, `UDS_SYS=UDS_SYS_ESP32`, and `UDS_SYS=UDS_SYS_ZEPHYR`, where `UDS_TP_ISOTP_C` is set by default.


CHANGELOG
## 0.11.0
- breaking API changes:
    - made client flags into bitfield:
| before                                       | after                                     |
| -----------------------------------------------------------------------------------------|
| `client->options |= UDS_SUPPRESS_POS_RESP`   | client->cfg_suppress_pos_resp = 1         |
| `client->options |= UDS_FUNCTIONAL`          | client->cfg_send_functional = 1           |
| `client->options |= UDS_IGNORE_SRV_TIMINGS`  | client->cfg_ignore_srv_sess_timing = 1    |


## 0.10.2
- revised release checklist

## 0.10.1
- fix vulnerabilities in 0x27 and 0x38 service handlers

## 0.10.0

- breaking API changes:
    - replaced Posix-specific ssize_t with UDSTpSize_t. 
    - replaced `UDSTpISOTpCInit` with server and client specific functions: `UDSServerTpISOTpCInit` and `UDSClientTpISOTpCInit`
- documentation overhauled

## 0.9.0
- breaking API changes:
    - converted subfunction enums to #defines with standard-consistent naming
    - simplified transport API
- refined release checklist in #60

## 0.8.0
- breaking API changes:
    - event enum consolidated `UDS_SRV_EVT_...` -> `UDS_EVT`
    - UDSClient refactored into event-based API
    - negative server response now raises a client error by default.
    - server NRCs prefixed with `UDS_NRC_`
    - NRCs merged into `UDS_Err` enum.
- added more examples of client usage


## 0.7.2
- runtime safety:
    1. turn off assertions by default, enable by `-DUDS_ENABLE_ASSERT`
    2. prefer `return UDS_ERR_INVALID_ARG;` over assertion in public functions
- use SimonCahill fork of isotp-c

## 0.7.1
- amalgamated sources into `iso14229.c` and `iso14229.h` to ease integration

## 0.7.0
- test refactoring. theme: test invariance across different transports and processor architectures
- breaking API changes:
    - overhauled transport layer implementation
    - simplified client and server init
    - `UDS_ARCH_` renamed to `UDS_SYS_`

## 0.6.0
- breaking API changes:
    - `UDSClientErr_t` merged into `UDSErr_t`
    - `TP_SEND_INPROGRESS` renamed to `UDS_TP_SEND_IN_PROGRESS`
    - refactored `UDSTp_t` to encourage struct inheritance
    - `UDS_TP_LINUX_SOCKET` renamed to `UDS_TP_ISOTP_SOCKET`
- added server fuzz test and qemu tests
- cleaned up example tests, added isotp-c on socketcan to examples
- added `UDS_EVT_DoScheduledReset`
- improve client error handling

## 0.5.0
- usability: refactored into a single .c/.h module
- usability: default transport layer configs are now built-in
- API cleanup: use `UDS` prefix on all exported functions
- API cleanup: use a single callback function for all server events

## 0.4.0
- refactor RDBIHandler to pass a function pointer that implements safe memmove rather than requiring the user to keep valid data around for an indefinite time or risking a buffer overflow.
- Prefer fixed-width. Avoid using `enum` types as return types and in structures.
- Transport layer is now pluggable and supports the linux kernel ISO-TP driver in addition to `isotp-c`. See [examples](./examples/README.md).

## 0.3.0
- added `iso14229ClientRunSequenceBlocking(...)`
- added server and client examples
- simplified test flow, deleted opaque macros and switch statements
- flattened client and server main structs
- simplified usage by moving isotp-c initialization parameters into server/client config structs 
- remove redundant buffers in server

## 0.2.0
- removed all instances of `__attribute__((packed))`
- refactored server download functional unit API to simplify testing
- refactored tests
    - ordered by service
    - documented macros
- removed middleware 
- simplified server routine control API
- removed redundant function `iso14229ServerEnableService`
- updated example

## 0.1.0
- Add client
- Add server SID 0x27 SecurityAccess
- API changes

## 0.0.0
- initial release



src/config.h
#pragma once

#include <assert.h> // early include for static_assert

/**
 * @def UDS_SYS
 * @brief Selects the host system iso14229 is compiled for.
 * @see uds_sys_ for the list of valid values
 */

/**
 * @def UDS_CUSTOM_MILLIS
 * @brief bring your own UDSMillis implementation
 * @details Bring your own UDSMillis implementation. Valid values:
 * - `0` (default): iso14229 provides UDSMillis() for the detected @ref UDS_SYS platform
 * - `1`: the user must provide their own UDSMillis() implementation
 *
 * @see UDSMillis
 */

/**
 * @def UDS_LOG_LEVEL
 * @brief sets the logging level
 * @details set UDS_LOG_LEVEL to 0U (UDS_LOG_NONE) to remove logging entirely.
 * @see uds_log_level_ for valid values
 */
#ifndef UDS_LOG_LEVEL
#define UDS_LOG_LEVEL 0U
#endif

#define UDS_ISOTP_MTU (4095U) ///< ISO-TP Maximum Transmission Unit (ISO-15764-2-2004 section 5.3.3)

#ifndef UDS_TP_MTU
/// ISOTP is the only supported tp type, so UDS inherits its MTU
#define UDS_TP_MTU UDS_ISOTP_MTU
#endif

/**
 * @def UDS_SERVER_SEND_BUF_SIZE
 * @brief reduce this at your own risk to save RAM. Fuzz testing is done with the default of @ref
 * UDS_TP_MTU.
 */
#ifndef UDS_SERVER_SEND_BUF_SIZE
#define UDS_SERVER_SEND_BUF_SIZE (UDS_TP_MTU)
#endif

/** @copydoc UDS_SERVER_SEND_BUF_SIZE */
#ifndef UDS_SERVER_RECV_BUF_SIZE
#define UDS_SERVER_RECV_BUF_SIZE (UDS_TP_MTU)
#endif

/** @copydoc UDS_SERVER_SEND_BUF_SIZE */
#ifndef UDS_CLIENT_SEND_BUF_SIZE
#define UDS_CLIENT_SEND_BUF_SIZE (UDS_TP_MTU)
#endif

/** @copydoc UDS_SERVER_SEND_BUF_SIZE */
#ifndef UDS_CLIENT_RECV_BUF_SIZE
#define UDS_CLIENT_RECV_BUF_SIZE (UDS_TP_MTU)
#endif

#ifndef UDS_CLIENT_DEFAULT_P2_MS
#define UDS_CLIENT_DEFAULT_P2_MS (150U) ///< default P2 timeout
#endif

#ifndef UDS_CLIENT_DEFAULT_P2_STAR_MS
#define UDS_CLIENT_DEFAULT_P2_STAR_MS (1500U) ///< default P2* timeout
#endif

static_assert(UDS_CLIENT_DEFAULT_P2_STAR_MS > UDS_CLIENT_DEFAULT_P2_MS, "");

#ifndef UDS_SERVER_DEFAULT_P2_MS
#define UDS_SERVER_DEFAULT_P2_MS (50) ///< default P2 duration
#endif

#ifndef UDS_SERVER_DEFAULT_P2_STAR_MS
#define UDS_SERVER_DEFAULT_P2_STAR_MS (5000) ///< default P2* duration
#endif

#ifndef UDS_SERVER_DEFAULT_S3_MS
#define UDS_SERVER_DEFAULT_S3_MS                                                                   \
    (5100) ///< default S3 duration (ISO14229-2 2013 Table 5: 5000 -0/+200 ms)
#endif

static_assert((0 < UDS_SERVER_DEFAULT_P2_MS) &&
                  (UDS_SERVER_DEFAULT_P2_MS < UDS_SERVER_DEFAULT_P2_STAR_MS) &&
                  (UDS_SERVER_DEFAULT_P2_STAR_MS < UDS_SERVER_DEFAULT_S3_MS),
              "");

/// Duration between the server sending a positive response to an ECU reset request and the emission
/// of a UDS_EVT_DoScheduledReset event. This should be set to a duration adequate for the server
/// transport layer to finish responding to the ECU reset request.
#ifndef UDS_SERVER_DEFAULT_POWER_DOWN_TIME_MS
#define UDS_SERVER_DEFAULT_POWER_DOWN_TIME_MS (60)
#endif

#if (UDS_SERVER_DEFAULT_POWER_DOWN_TIME_MS < UDS_SERVER_DEFAULT_P2_MS)
#error "The server shall have adequate time to respond before reset"
#endif

/// Amount of time to wait after boot before accepting 0x27 requests.
#ifndef UDS_SERVER_0x27_BRUTE_FORCE_MITIGATION_BOOT_DELAY_MS
#define UDS_SERVER_0x27_BRUTE_FORCE_MITIGATION_BOOT_DELAY_MS (1000U)
#endif

/// Amount of time to wait after an authentication failure before accepting another 0x27 request.
#ifndef UDS_SERVER_0x27_BRUTE_FORCE_MITIGATION_AUTH_FAIL_DELAY_MS
#define UDS_SERVER_0x27_BRUTE_FORCE_MITIGATION_AUTH_FAIL_DELAY_MS (1000U)
#endif

#ifndef UDS_SERVER_DEFAULT_XFER_DATA_MAX_BLOCKLENGTH
/*! ISO14229-1:2013 Table 396. This parameter is used by the requestDownload positive response
message to inform the client how many data bytes (maxNumberOfBlockLength) to include in each
TransferData request message from the client. */
#define UDS_SERVER_DEFAULT_XFER_DATA_MAX_BLOCKLENGTH (UDS_TP_MTU)
#endif

#ifndef UDS_AUTOSELECT_TP // transport auto-selection. Set -DUDS_AUTOSELECT_TP=0 to turn off
                          // auto-selection
#define UDS_AUTOSELECT_TP 1

#if ((UDS_SYS == UDS_SYS_CUSTOM) || (UDS_SYS == UDS_SYS_ARDUINO) || (UDS_SYS == UDS_SYS_ESP32) ||  \
     (UDS_SYS == UDS_SYS_ZEPHYR))
#ifndef UDS_TP_ISOTP_C
#define UDS_TP_ISOTP_C
#endif // #ifndef UDS_TP_ISOTP_C
#elif (UDS_SYS == UDS_SYS_UNIX)
#ifndef UDS_TP_ISOTP_SOCK
#define UDS_TP_ISOTP_SOCK
#endif // #ifndef UDS_TP_ISOTP_SOCK
#else
// no default tp
#endif // ((UDS_SYS == UDS_SYS_CUSTOM) || ... )

#endif // #ifndef UDS_AUTOSELECT_TP
